pkg/p2p/server.go

```diff
@@ -0,0 +1,844 @@
+}
+
+type taskOnMessageBatch struct {
+	streamMeta     *p2p.StreamMeta
+	messageEntries []*p2p.MessageEntry
+}
+
+type taskOnRegisterPeer struct {
+	sender     *streamHandle
+	clientAddr string // for logging
+}
+
+type taskOnRegisterHandler struct {
+	topic   string
+	handler workerpool.EventHandle
+	done    chan struct{}
+}
+
+type taskOnDeregisterHandler struct {
+	topic string
+	done  chan struct{}
+}
+
+// taskDebugDelay is used in unit tests to artificially block the main
+// goroutine of the server. It is not used in other places.
+type taskDebugDelay struct {
+	doneCh chan struct{}
+}
+
+// NewMessageServer creates a new MessageServer
+func NewMessageServer(serverID NodeID, config *MessageServerConfig) *MessageServer {
+	return &MessageServer{
+		serverID:        serverID,
+		handlers:        make(map[string]workerpool.EventHandle),
+		peers:           make(map[string]*cdcPeer),
+		pendingMessages: make(map[topicSenderPair][]pendingMessageEntry),
+		acksMap:         make(map[NodeID]map[Topic]Seq),
+		taskQueue:       make(chan interface{}, config.MaxPendingTaskCount),
+		pool:            workerpool.NewDefaultWorkerPool(config.WorkerPoolSize),
+		closeCh:         make(chan struct{}),
+		config:          config,
+	}
+}
+
+// Run starts the MessageServer's worker goroutines.
+// It must be running to provide the gRPC service.
+func (m *MessageServer) Run(ctx context.Context) error {
+	atomic.StoreInt32(&m.isRunning, 1)
+	defer func() {
+		atomic.StoreInt32(&m.isRunning, 0)
+		close(m.closeCh)
+	}()
+
+	errg, ctx := errgroup.WithContext(ctx)
+	errg.Go(func() error {
+		return errors.Trace(m.run(ctx))
+	})
+
+	errg.Go(func() error {
+		return errors.Trace(m.pool.Run(ctx))
+	})
+
+	return errg.Wait()
+}
+
+func (m *MessageServer) run(ctx context.Context) error {
+	ticker := time.NewTicker(m.config.AckInterval)
+	defer ticker.Stop()
+
+	for {
+		failpoint.Inject("ServerInjectTaskDelay", func() {
+			log.Info("channel size", zap.Int("len", len(m.taskQueue)))
+		})
+		select {
+		case <-ctx.Done():
+			return errors.Trace(ctx.Err())
+		case <-ticker.C:
+			m.tick(ctx)
+		case task := <-m.taskQueue:
+			switch task := task.(type) {
+			case taskOnMessageBatch:
+				for _, entry := range task.messageEntries {
+					if err := m.handleMessage(ctx, task.streamMeta, entry); err != nil {
+						return errors.Trace(err)
+					}
+				}
+			case taskOnRegisterHandler:
+				// TODO think about error handling here
+				if err := m.registerHandler(ctx, task.topic, task.handler, task.done); err != nil {
+					return errors.Trace(err)
+				}
+				log.Debug("handler registered", zap.String("topic", task.topic))
+			case taskOnDeregisterHandler:
+				if handler, ok := m.handlers[task.topic]; ok {
+					delete(m.handlers, task.topic)
+					go func() {
+						err := handler.GracefulUnregister(ctx, time.Second*5)
+						if err != nil {
+							log.L().DPanic("failed to gracefully unregister handle",
+								zap.Error(err))
+						}
```

Comment:
In which scenario will `GracefulUnregister` return error, is the underlying work pool blocked? What will happen if don't panic here

Reply:
`GracefulUnregister` will return error if the `ctx` is cancelled or the timeout has passed. `DPanic` here is for debugging and will not have an effect in production.

The underlying workerpool is not blocked in anyway.